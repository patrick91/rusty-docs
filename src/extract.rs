use crate::docstrings;
use rustpython_ast::{Constant, ExprKind, StmtKind, Located};
use rustpython_parser::parser;

pub struct Argument {
    pub name: String,
    pub type_: String,
}

#[derive()]
pub struct Function {
    pub name: String,
    // TODO: arguments?
    pub docstring: docstrings::Docstring,
    pub arguments: Vec<Argument>,
}

pub struct Module {
    // pub name: String,
    // pub docstring: docstrings::Docstring,
    pub functions: Vec<Function>,
}

fn extract_function(name: String, body: Vec<Located<StmtKind>>) -> Function {
    // find docstring, the first statement in the body of the function
    // that's an Expr with a Constant value

    let docstring_text = match body.first() {
        Some(stmt) => match &stmt.node {
            StmtKind::Expr { value } => match &value.node {
                ExprKind::Constant { value, kind: _ } => match value {
                    Constant::Str(value) => value.clone(),
                    _ => "".to_string(),
                },
                _ => "".to_string(),
            },
            _ => "".to_string(),
        },
        None => "".to_string(),
    };

    let docstring = docstrings::Docstring::new_from_string(&docstring_text);
    let arguments = Vec::new();

    Function {
        name: name.to_string(),
        docstring,
        arguments,
    }
}

pub fn extract(code: &str) -> Module {
    let python_ast = parser::parse_program(&code, "something").expect("Unable to parse");

    let mut functions = Vec::new();

    // find all functions in ast
    for statement in python_ast {
        match statement.node {
            StmtKind::FunctionDef {
                name,
                body,
                args,
                decorator_list,
                returns,
                type_comment,
            }=> {
                functions.push(extract_function(name, body));
            }
            _ => {}
        }
    }

    Module { functions }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    #[test]
    fn it_parses_python_function() {
        let code = r#"
        def foo(a: int, b: int) -> int:
            """Example docstring

            Args:
                a: a number
                b: another number

            Returns:
                the sum of a and b

            Raises:
                ValueError: if a or b are not numbers
            """
            return a + b
        "#;

        let result = super::extract(&textwrap::dedent(code));

        assert_eq!(result.functions.len(), 1);

        let function = &result.functions[0];

        assert_eq!(function.name, "foo");
        assert_eq!(function.docstring.title, "Example docstring");
        assert_eq!(function.docstring.description, "");
        assert_eq!(function.docstring.arguments.len(), 2);

        let arguments = &function.docstring.arguments;

        assert_eq!(arguments[0].name, "a");
        assert_eq!(arguments[0].description, Some("a number".to_string()));
        assert_eq!(arguments[1].name, "b");
        assert_eq!(arguments[1].description, Some("another number".to_string()));

        assert_eq!(function.docstring.returns, "the sum of a and b");
        assert_eq!(function.docstring.raises.len(), 1);
        assert_eq!(function.docstring.raises[0].exception, "ValueError");
        assert_eq!(
            function.docstring.raises[0].description,
            Some("if a or b are not numbers".to_string())
        );
    }
}
