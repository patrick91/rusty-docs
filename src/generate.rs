use crate::extract;
use std::{
    fs,
    path::{Component, Path, PathBuf},
};

use minijinja::{context, Environment};

pub fn normalize_path(path: &Path) -> PathBuf {
    let mut components = path.components().peekable();
    let mut ret = if let Some(c @ Component::Prefix(..)) = components.peek().cloned() {
        components.next();
        PathBuf::from(c.as_os_str())
    } else {
        PathBuf::new()
    };

    for component in components {
        match component {
            Component::Prefix(..) => unreachable!(),
            Component::RootDir => {
                ret.push(component.as_os_str());
            }
            Component::CurDir => {}
            Component::ParentDir => {
                ret.pop();
            }
            Component::Normal(c) => {
                ret.push(c);
            }
        }
    }
    ret
}

// multline string
const TEMPLATE: &str = r#"
# {{ function_name }}

{{ function_docstring.title }}
{% for part in function_docstring.body -%}
{%-if part.CodeSnippet -%}
```python
{{ part.CodeSnippet }}
```
{%-elif part.Text %}
{{ part.Text }}
{% endif %}
{% endfor-%}

{% if function_docstring.arguments %}
## Arguments:


| Name | Type | Description |
| --- | --- | --- |
{% for argument in function_arguments-%}
| {{ argument.name }} | {{ argument.type }} | {{ argument.description }} |
{% endfor %}
{% endif %}

{% if function_docstring.returns %}
## Returns:

{{ function_docstring.returns }}
{% endif %}

{% if function_docstring.raises %}
## Raises:

{{ function_docstring.raises }}
{% endif %}
"#;

pub fn generate(code: &str) -> String {
    let module = extract::extract(code);

    // let path = Path::new("./templates/function.md");

    // let path = if path.exists() {
    //     path
    // } else {
    //     Path::new("./node_modules/rusty_docs/templates/function.md")
    // };

    // println!("normalised path is: {:?}", normalize_path(path));

    let function_template = TEMPLATE; // fs::read_to_string(path).expect("Unable to read function template");

    let mut env = Environment::new();
    env.add_template("function", &function_template).unwrap();

    let function_template = env.get_template("function").unwrap();

    let mut output = String::new();

    for function in module.functions {
        let content = function_template
            .render(context!(
                function_name => function.name,
                function_docstring => function.docstring,
                function_arguments => function.arguments
            ))
            .unwrap();

        output.push_str(&content);
    }

    output
}

// #[cfg(test)]
// mod tests {

//     #[cfg(test)]
//     glob!("./src/fixtures/*.py", |path| {
//         let code = fs::read_to_string(path).unwrap();

//         insta::assert_display_snapshot!(super::generate(&code));
//     });
// }

#[test]
fn test_snapshots() {
    insta::glob!("./src/fixtures/**/**/.py", |path| {
        let contents = std::fs::read_to_string(path).unwrap();
        insta::assert_display_snapshot!(&contents);
    });
}
// use insta::glob;

// #[cfg(test)]
// glob!("./src/fixtures/*.py", |path| {
//     let code = fs::read_to_string(path).unwrap();

//     insta::assert_display_snapshot!(super::generate(&code));
// });
