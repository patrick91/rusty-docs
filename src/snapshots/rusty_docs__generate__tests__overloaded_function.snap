---
source: src/generate.rs
expression: "super::generate(&code)"
---

# f







# f







# f

This is a docstring

## Arguments:


| Name | Type | Description |
| --- | --- | --- |
| x | str | int | This is a parameter |




## Returns:

This is a return value



